### PR TITLE
Issue #1300 by docwilmot: Add ability to add classes to layout regions.

### PR DESCRIPTION
**Pull request by jenlampton**
PR for https://github.com/backdrop/backdrop-issues/issues/1300

### CODE REVIEW

File: core/modules/layout/layout.admin.inc [hunk 1]
```diff
@@ -2297,3 +2297,73 @@ function layout_ajax_form_update($form, $form_state) {
...
+  );
+
+  return $form;
+}
+ 
+/**
+ * AJAX callback to update the style settings within the region styles form.
+ */
+function layout_region_configure_ajax_style($form, $form_state) {
+  return $form;
```
Comment (docwilmot) [inline core/modules/layout/layout.admin.inc]: Not needed anymore

File: core/modules/layout/layout.module [hunk 2]
```diff
@@ -653,6 +646,7 @@ function layout_autoload_info() {
     // Styles.
     'LayoutStyle' => 'plugins/styles/layout_style_default.inc',
     'LayoutStyleDynamic' => 'plugins/styles/layout_style_dynamic.inc',
+    'LayoutStyleRegion' => 'plugins/styles/layout_style_region.inc',
```
Comment (docwilmot) [inline core/modules/layout/layout.module]: Not needed anymore

File: core/modules/layout/layout.admin.inc (hunk 1 above)
Comment (jenlampton) [inline core/modules/layout/layout.admin.inc]: pushed a fix for this.

File: core/modules/layout/layout.module (hunk 2 above)
Comment (jenlampton) [inline core/modules/layout/layout.module]: pushed a fix for this too.  good catch!

File: core/modules/layout/layout.admin.inc [hunk 3]
```diff
@@ -2297,3 +2297,66 @@ function layout_ajax_form_update($form, $form_state) {
...
+  if (!array_key_exists($region_name, $layout_info['regions'])) {
+    return MENU_NOT_FOUND;
+  }
+
+  $form['#tree'] = TRUE;
+
+  $options = array(
+    '' => 'No inner wrapper',
+    'div' => 'DIV',
+    'navgit' => 'NAV',
```
Comment (quicksketch) [inline core/modules/layout/layout.admin.inc]: `navgit` => `nav`.

I suppose it's possible that an entire region could be a `nav`, e.g. a mega-menu type situation where the menu could be made up of multiple blocks, like both a menu and a view.

File: core/modules/layout/layout.api.php [hunk 4]
```diff
@@ -140,12 +137,11 @@ function hook_layout_context_info() {
  *   - file: Optional. If using theme or preprocess functions, specify the path
  *     to the containing file. This file path is NOT the path to the class.
  *     Class paths and loading is done through hook_autoload_info().
- *   - path: The relative path from this module to the directory containing any
+ *   - template: The relative path from this module to the directory containing any
```
Comment (quicksketch) [inline core/modules/layout/layout.api.php]: `path` is still a valid key. See `layout_theme()` for where it is used. So it should be left as-is. We should add a separate item for `template` though, which isn't currently documented.

File: core/modules/layout/layout.api.php [hunk 5]
```diff
@@ -155,16 +151,14 @@ function hook_layout_style_info() {
...
-    'region theme' => 'mymodule_layout_region',
     // The theme key for rendering an individual block.
     'block theme' => 'mymodule_block',
     // Provide a class name if this style has settings. The class should extend
     // the LayoutStyle class.
     'class' => 'MyModuleLayoutStyle',
     // If nesting templates or a theme include file in a directory, the relative
     // path can be specified.
-    'path' => 'templates',
+    'template' => 'templates/mymodule-file.tpl.php',
```
Comment (quicksketch) [inline core/modules/layout/layout.api.php]: Revert this, `path` is still valid. Add a separate entry for `template` if desired.

File: core/modules/layout/layout.theme.inc [hunk 6]
```diff
@@ -137,6 +137,19 @@ function theme_layout_conditions($variables) {
...
 
 /**
+ * Output region content wrapped by tag specified in layout UI.
+ */
+function theme_layout_region_inner($variables) {
+  $tag = $variables['tag'];
+  $classes = $variables['classes'];
+
+  if ($tag != '') {
+    return '<' . $tag . ' class="' . $classes .'">' . implode('', $variables['blocks']) . '</' . $tag . '>';
```
Comment (quicksketch) [inline core/modules/layout/layout.theme.inc]: For flexibility and consistency with block and layout templates, $attributes should also be passed in and added to the wrapper.

File: core/modules/layout/layout.admin.inc (hunk 3 above)
Comment (jenlampton) [inline core/modules/layout/layout.admin.inc]: fixed.

File: core/modules/layout/layout.api.php (hunk 4 above)
Comment (jenlampton) [inline core/modules/layout/layout.api.php]: updated.

File: core/modules/layout/layout.api.php (hunk 5 above)
Comment (jenlampton) [inline core/modules/layout/layout.api.php]: fixed.

File: core/modules/layout/layout.theme.inc (hunk 6 above)
Comment (jenlampton) [inline core/modules/layout/layout.theme.inc]: There aren't any attributes here.  Should I create an array in the theme function?  Something like this...

``` php
function theme_layout_region_inner($variables) {
  $tag = $variables['tag'];
  $attributes = array('class' => explode($variables['classes']));

  if ($tag != '') {
    return '<' . $tag . ' ' . backdrop_attributes($attributes) .'>' . implode('', $variables['blocks']) . '</' . $tag . '>';
  }
  return implode('', $variables['blocks']);
}
```

**Comment by quicksketch** [comment]
Looks good but we need to have the failing test fixed.

**Comment by jenlampton** [comment]
Fixed the failing test.

**Comment by quicksketch** [comment]
Merged into 1.x for 1.4.0. Thanks @docwilmot and @jenlampton!

**Pull request merged**